hasher-matcher-actioner/hmalib/common/mappings.py

```diff
@@ -1,37 +1,284 @@
+    FBThreatExchangeSignalExchangeAPI,
+)
+from threatexchange.content_type.photo import PhotoContent
+from threatexchange.content_type.video import VideoContent
+from threatexchange.signal_type.md5 import VideoMD5Signal
+from threatexchange.signal_type.pdq import PdqSignal
 
-from hmalib.indexers.s3_indexers import (
-    S3BackedMD5Index,
-    S3BackedPDQIndex,
-    S3BackedPDQFlatIndex,
-    S3BackedInstrumentedIndexMixin,
+from hmalib.common.config import (
+    HMAConfig,
+    create_config,
+    create_or_update_config,
+    update_config,
 )
+from hmalib.common.logging import get_logger
+from hmalib.aws_secrets import AWSSecrets
+
+# Used if this is the first time HMA is being booted. TODO: Should this be
+# configurable before boot?
+DEFAULT_SIGNAL_AND_CONTENT_TYPES = SignalTypeMapping(
+    content_types=[PhotoContent, VideoContent], signal_types=[VideoMD5Signal, PdqSignal]
+)
+
+logger = get_logger(__name__)
+
+
+def import_class(full_class_name) -> t.Type:
+    module_name, klass = full_class_name.rsplit(".", 1)
+    module = importlib.import_module(module_name)
+    return getattr(module, klass)
+
+
+"""
+Reverses import_class. Will create a module.ClassName style string that can
+be imported using import_class.
+"""
+
+
+def full_class_name(klass) -> str:
+    return f"{klass.__module__}.{klass.__name__}"
+
+
+@dataclass
+class ToggleableSignalTypeConfig(HMAConfig):
+    signal_type_class: str
+
+    # Allow soft disables.
+    enabled: bool = True
+
+    @staticmethod
+    def get_name_from_type(signal_type: t.Type[SignalType]) -> str:
+        return f"signal_type:{full_class_name(signal_type)}"
+
+
+@dataclass
+class ToggleableContentTypeConfig(HMAConfig):
+    content_type_class: str
+
+    # Allow soft disables.
+    enabled: bool = True
+
+    @staticmethod
+    def get_name_from_type(content_type: t.Type[ContentType]) -> str:
+        return f"content_type:{full_class_name(content_type)}"
+
+
+class HMASignalTypeMapping(SignalTypeMapping):
+    """
+    An extension of the SignalTypeMapping defined in threatexchange.meta.
+
+    Enhancements:
+
+    1. State is pulled in from dynamodb when __init__ is first called.
+    2. Dedicated methods for getting signal and content_type by name. Instead of
+       depending on SignalTypeMapping's internal state.
+    """
+
+    @classmethod
+    def get_from_config_or_default(cls) -> "HMASignalTypeMapping":
+        """
+        Pull configs from HMAConfigs if available. If not, use defaults.
+        """
+        all_content_types = ToggleableContentTypeConfig.get_all()
+        all_signal_types = ToggleableSignalTypeConfig.get_all()
+
+        if len(all_content_types) + len(all_signal_types) == 0:
+            # We have never written content or signal types to database, so use
+            # default values instead.
+            (
+                all_content_types,
+                all_signal_types,
+            ) = cls._get_default_configs()
```

Comment:
blocking: This means as soon as they add the first st/ct config, they may suddenly have an unexpected change in configuration because it erases the others.

I'd suggest instead this behavior:

```
all_content_types, all_signal_types = cls._get_default_configs()
by_class = {ct.content_type_class: ct for ct in all_content_types}
by_class.update(ct.content_type_class: ct for ct in ToggleableContentTypeConfig.get_all()}
all_content_types = list(by_class.values())
# Same for signal
```

Basically, implicitly, these configs exist until they are inserted into the database, in which case the database version exists.

This creates unusual behavior when you delete these records unfortunately, where disabled => deleted turns into enabled. To fix would require us to prevent deletion from the UI for these, and instead turn into disable (which I think makes sense, but want your input).

Reply:
I believe we discussed this to death on call. The default set of records need to be handled, and is not handled at the moment. 

Solution we agreed on is close to (exactly?) what you suggested. I will put it on top of this stack.